### PR TITLE
GeoMx mask generation

### PR DESCRIPTION
**Pull request by lidanwu**
As stated in issue #63, this PR introduces a python script that does automatic binary mask generation in batch for single-cell application with respect to query marker stain in neighborhood. This could be useful in neuropathology studies.

### CODE REVIEW

File: _code/GeoMx-scMask-generation/Generate_scMask_given_query.py [hunk 1]
```diff
@@ -0,0 +1,372 @@
+"""
+Copyright © 2024 NanoString Technologies (a Bruker company), Authored by Lidan Wu.
```
Comment (eveilyeverafter) [inline _code/GeoMx-scMask-generation/Generate_scMask_given_query.py]: Can you confirm this?

See [here](https://github.com/Nanostring-Biostats/CosMx-Analysis-Scratch-Space/blob/02bc8fd58ee8890546659568b998827674401a67/assets/make-composite/make_composite.py#L1) for example of a recently used header.

File: posts/GeoMx-scMask-generation/index.qmd [hunk 2]
```diff
@@ -0,0 +1,167 @@
...
+image: figures/workflow_GeoMx_scMask_given_query.png
+description: "This post introduces a pipeline for automatic generating GeoMx&#8482;-ready binary masks in batch for marker-based single-cell application. Given the query marker protein of interest, the pipeline would take morphology images and generate binary masks for negative-stained cells and cells connecting to positive-stained regions, respectively. The pipeline runs as a command line and this post would serve as a guide to how it works and how to use it."
+draft: false
+execute:
+  engine: knitr
+---
+
+# Introduction
+
+Ability to separate cells based on their relationship to nearby marker protein staining is a desired feature in many biological studies. On such use case would be to study the difference between neurons immediately next to neurofibrillary tangles and neurons that are a little further away. This post introduces a pipeline for automatic generating GeoMx&#8482;-ready binary masks in batch for marker-based single-cell application.
```
Comment (eveilyeverafter) [inline posts/GeoMx-scMask-generation/index.qmd]: Ability -> The ability.

automatic -> automatically 

Add Digital Spatial Profiler (DSP) after GeoMx. i.e., automatically generating GeoMx Digital Spatial Profiler (DSP)-ready binary masks...

File: posts/GeoMx-scMask-generation/index.qmd [hunk 3]
```diff
@@ -0,0 +1,167 @@
...
+
+Like other items in our [Analysis Scratch Space](https://nanostring-biostats.github.io/CosMx-Analysis-Scratch-Space/about.html){target="_blank"},
+the usual [caveats](https://nanostring-biostats.github.io/CosMx-Analysis-Scratch-Space/about.html){target="_blank"} and [license](https://nanostring-biostats.github.io/CosMx-Analysis-Scratch-Space/license.html){target="_blank"} applies.
+
+
+
+# Workflow Description
+
+The pipeline is a python script `Generate_scMask_given_query.py` that runs as a command line and you can downloaded from `_code/GeoMx-scMask-generation` [folder](https://github.com/Nanostring-Biostats/CosMx-Analysis-Scratch-Space/tree/Main/_code/GeoMx-scMask-generation){target="_blank"} in this repository. By passing file path to image folder to the script, the pipeline would take GeoMx&#8482; morphology images and generate binary masks for negative-stained cells and cells connecting to positive-stained regions, respectively.
```
Comment (eveilyeverafter) [inline posts/GeoMx-scMask-generation/index.qmd]: passing _the_ file path

File: posts/GeoMx-scMask-generation/index.qmd [hunk 4]
```diff
@@ -0,0 +1,167 @@
...
+Like other items in our [Analysis Scratch Space](https://nanostring-biostats.github.io/CosMx-Analysis-Scratch-Space/about.html){target="_blank"},
+the usual [caveats](https://nanostring-biostats.github.io/CosMx-Analysis-Scratch-Space/about.html){target="_blank"} and [license](https://nanostring-biostats.github.io/CosMx-Analysis-Scratch-Space/license.html){target="_blank"} applies.
+
+
+
+# Workflow Description
+
+The pipeline is a python script `Generate_scMask_given_query.py` that runs as a command line and you can downloaded from `_code/GeoMx-scMask-generation` [folder](https://github.com/Nanostring-Biostats/CosMx-Analysis-Scratch-Space/tree/Main/_code/GeoMx-scMask-generation){target="_blank"} in this repository. By passing file path to image folder to the script, the pipeline would take GeoMx&#8482; morphology images and generate binary masks for negative-stained cells and cells connecting to positive-stained regions, respectively. 
+
+![Workflow schematic](figures/workflow_GeoMx_scMask_given_query.png){#fig-workflow}
```
Comment (eveilyeverafter) [inline posts/GeoMx-scMask-generation/index.qmd]: Can you be more descriptive here so that someone with a neuro background can follow. For example, what are Cyto and Query staining. I'm having trouble relating this figure to the initial example in the introduction.

File: posts/GeoMx-scMask-generation/index.qmd [hunk 5]
```diff
@@ -0,0 +1,167 @@
...
+    + The rearranged images would be saved in `2chan_res` (or `3chan_res` if including nuclear stain as 3rd channel) sub folder of the user-designated output root folder. 
+2. After rearrangement, the pipeline script would perform cell segmentation using cytoplasm stain and optionally nuclear stain as well. The resulting cell label images with pixel value indicates cell ID would be saved under `cellLabels` sub folder within the step (1) result folder.
+3. The pipeline script also takes the query marker stain of each image and performs auto-threshold to identify individual objects with positive stain and **bigger** than user-defined size cutoff, `min_positive_area`.
+4. With label images from both cell segmentation and query-stained object segmentation, the pipeline script would then perform connectivity analysis between the 2 sets of label images and output 2 sets of binary masks in separate folders under user-designated output root folder. 
+    i) `negative_cell_masks`: binary masks for cells that does **NOT** have valid overlapping with any of positive-stained object.
```
Comment (eveilyeverafter) [inline posts/GeoMx-scMask-generation/index.qmd]: cells that _do_ **NOT**

File: posts/GeoMx-scMask-generation/index.qmd [hunk 6]
```diff
@@ -0,0 +1,167 @@
...
+3. The pipeline script also takes the query marker stain of each image and performs auto-threshold to identify individual objects with positive stain and **bigger** than user-defined size cutoff, `min_positive_area`.
+4. With label images from both cell segmentation and query-stained object segmentation, the pipeline script would then perform connectivity analysis between the 2 sets of label images and output 2 sets of binary masks in separate folders under user-designated output root folder. 
+    i) `negative_cell_masks`: binary masks for cells that does **NOT** have valid overlapping with any of positive-stained object. 
+         +  A valid overlapping is defined by whether the intersected area between cell and positive-stained object is **bigger** than user-defined size cutoff, `min_intersect_area`. 
+    ii) `positive_ROI_masks`: binary masks for cells with valid overlapping event **AND** their connected positive-stained objects. 
+
+At this point, user can upload the output binary masks back to GeoMx&#8482; DSP instrument for the corresponding ROIs and initiate the sample collection.
```
Comment (eveilyeverafter) [inline posts/GeoMx-scMask-generation/index.qmd]: I think this would be good to put higher up to help explain _why_ a user may want to take this approach.

Comment (eveilyeverafter) [inline posts/GeoMx-scMask-generation/index.qmd]: I would reword this sentence...

Once complete, users can upload these binary masks back into the DSP instrument for corresponding ROI selection.

File: posts/GeoMx-scMask-generation/index.qmd [hunk 7]
```diff
@@ -0,0 +1,167 @@
...
+
+```{default filename="Command Line Interface (CLI) example to run in Terminal"}
+#| eval: false
+python Generate_scMask_given_query.py --in_dir imageFolder --cyto_chan 4 --query_chan 3 --out_dir outputFolder --min_positive_area 500
+```
+
+The resulting output folder structure looks like this:
+
+```
+outputFolder
```
Comment (eveilyeverafter) [inline posts/GeoMx-scMask-generation/index.qmd:98]: This is really helpful. Thanks for adding this.

**Review by eveilyeverafter — APPROVED** [review]
Short post and minor edits requested. General comment: I think it would help the readers to have a clearer example of why this method is useful within the first few sentences of the post (since this is not our usual CosMx post). 

I did not review the code itself extensively has I'm sure you've been using and testing it for a while.

I don't need to re-review it after changes are made. Feel free to resolve individual comments as required.

File: posts/GeoMx-scMask-generation/index.qmd (hunk 4 above)
Comment (lidanwu) [reply posts/GeoMx-scMask-generation/index.qmd]: I added a more descriptive caption for the figure and now it reads:

> Workflow schematic. In neuropathology study, it's sometimes desired to separate between neurons (stained by NeuN in red channel) touching the neurofibrillary tangles (stained by Phospho-Tau in green channel) and those that are a little further away. The pipeline described here fullfills the need of separating cells based on additional morphology marker of interest (aka "Query") and could also be applied to other single-cell applications like separating immune cells based on their contact with nearby pathogen.

File: posts/GeoMx-scMask-generation/index.qmd (hunk 6 above)
Comment (lidanwu) [reply posts/GeoMx-scMask-generation/index.qmd]: I'm not following. It says in the 1st paragraph that this post is about how to generate GeoMx masks. Could you please clarify how would you like it changed?

Comment (lidanwu) [reply posts/GeoMx-scMask-generation/index.qmd]: ROI in terms of sample area on tissue is already chosen and that's where user downloads the morphology images in first place, so no further selection. The masks are to further refine the sample collection into different zones. It might be confusing to say ROI selection again. But I have adopted your suggestions with minor edits.

Comment (lidanwu) [reply posts/GeoMx-scMask-generation/index.qmd]: In the workflow step (3) and (4.i), I added additional bullet points explain how the size cutoff  would help.

step(3)
> You can adjust the size cutoff based on your particular application to get a cleaner isolation of objects of interest. For example, if you are only interested in mature tangles, you can use higher size cutoff to select only the query-positive objects of big size. 

step(4.i)

> The ability to define valid overlapping  based on area cutoff could help to separate between barely touching events and more significant overlapping events. This could be useful if you want to separate between immune cells just contacting a microbe stained in query channel and immune cells that have engulfed an invading microbe.
